File: examples/python-sdk/ex07_permissions.py
"""ex07 — Human-in-the-loop tool approval (on_permission). ONE example, two
transports:

    python ex07_permissions.py ipc
    python ex07_permissions.py in_process

`jaato.session(mode=...)` picks the transport (the daemon via IPC, or the
embedded in-process runtime); the SAME `profile={...}` and kwargs run both ways.
`socket_path` is ipc-only (ignored in-process); `env_file` applies to both.

Appears in: pydantic-ai.md §7 (canonical/majority form), agno.md §7,
claude-agent.md §7, openai-agents.md §7, strands.md §7, langchain.md §7*.
(* langchain.md uses a named `def approve(ev)` that prompts via `input()`; the
five other Python docs share this inline-lambda form, so it is canonical.)

The doc leaves `approve(tool_name)` as an illustrative predicate. Here it is a
real one — auto-approving so the example is headless (the langchain variant
prompts via input()).

The on_permission mechanism is identical for any gated tool. The prompt asks for
something only the shell can provide (the system `date`), and `defaultPolicy:"ask"`
gates every cli call → on_permission is always asked → approves. This makes the
gate fire deterministically rather than the doc's "Delete temp.log", which a model
may decline or answer conversationally without calling the tool.

Standing deviations (see README): the dedicated-daemon connection, the
model/provider literal + `**AUTH` (pass: cred knob), the ask-policy + benign
command above.
"""
import asyncio
import sys

import jaato
from _config import CONN, WORKSPACE, AUTH

mode = sys.argv[1] if len(sys.argv) > 1 else "ipc"


def approve(tool_name: str) -> bool:
    # In a TUI this would prompt; headless, we auto-approve and log the gate.
    print(f"[permission] {tool_name} -> approve")
    return True


async def main():
    async with jaato.session(
        mode=mode,
        profile={"model": "google/gemini-2.5-flash", "provider": "openrouter", "plugins": ["cli(preload)"],
                 "suppress_base_instructions": True,
                 "plugin_configs": {**AUTH["plugin_configs"],
                                    "permission": {"policy": {"defaultPolicy": "ask"}}}},
        env_file=CONN["env_file"],
        socket_path=CONN["socket_path"],   # ipc-only; ignored in-process
        workspace_path=WORKSPACE,
        on_permission=lambda ev: "y" if approve(ev.tool_name) else "n",
    ) as s:
        print(f"[{mode}]", await s.ask("What is the current date and time on this machine? Find out by running the `date` command in the shell."))


asyncio.run(main())

File: examples/ts-sdk/src/ex07_permissions.ts
// ex07 — Human-in-the-loop tool approval (onPermission).
//
// Appears in: mastra.md §7.
//
// Doc snippet (verbatim shape):
//
//   await using s = await JaatoClient.session({
//     url, profile: { model: "gpt-4o", provider: "openai", plugins: ["cli"] },
//     onPermission: (ev) => (confirm(`allow ${ev.tool_name}?`) ? "y" : "n"),
//   });
//   console.log(await s.ask("Delete temp.log"));
//
// Same finding as examples/python-sdk/ex07: the doc's "Delete temp.log" is a
// flaky gate trigger (models refuse a destructive shell delete, or answer
// conversationally instead of calling the tool). The onPermission mechanism is
// identical for ANY gated tool, so we make the gate fire deterministically: ask
// for something only the shell can provide (the system `date`) with
// defaultPolicy "ask" so every cli call is gated → onPermission is asked.
//
// Standing substitutions (see README): `...CONN`, `...AUTH`, the model literal,
// the ask-policy + benign command.

import { JaatoClient } from "@jaato/sdk";
import { CONN, AUTH, WORKSPACE, CONFIG_ROOT } from "./_config.js";

function approve(toolName: string): boolean {
  // In a UI this would prompt; headless, we auto-approve and log the gate.
  console.log(`[permission] ${toolName} -> approve`);
  return true;
}

await using s = await JaatoClient.session({
  ...CONN,
  workspacePath: WORKSPACE,
  configRoot: CONFIG_ROOT,
  profile: {
    model: "google/gemini-2.5-flash",
    provider: "openrouter",
    plugins: ["cli"],
    plugin_configs: { ...AUTH.plugin_configs, permission: { policy: { defaultPolicy: "ask" } } },
  },
  onPermission: (ev) => (approve((ev as { tool_name: string }).tool_name) ? "y" : "n"),
});
console.log(
  await s.ask("What is the current date and time on this machine? Find out by running the `date` command in the shell."),
);
